File: golang/conway.go
package main

import (
	"fmt"
	"io/ioutil"
	"log"
	"os"
	"strconv"
	"strings"
)

type CellState uint

const (
	Dead  CellState = 0
	Alive CellState = 1
)

func (c CellState) String() string {
	switch c {
	case Dead:
		return "."
	case Alive:
		return "#"
	default:
		return fmt.Sprintf("%d", int(c))
	}
}

type ConwayGrid struct {
	width  uint
	height uint
	data   [][]CellState
}

func (cg ConwayGrid) String() string {
	var stringGird string
	for _, row := range cg.data {
		for _, cell := range row {
			stringGird += fmt.Sprintf("%s", cell)
		}
		stringGird += "\n"
	}
	return stringGird
}

func (cg ConwayGrid) step() {
	oldGridData := make([][]CellState, len(cg.data), len(cg.data[0]))
	copy(oldGridData, cg.data)
	oldGrid := ConwayGrid{cg.width, cg.height, oldGridData}
	for r, _ := range cg.data {
		for c, _ := range cg.data[r] {
			cg.updateCell(r, c, &oldGrid)
		}
	}
}

func (cg ConwayGrid) updateCell(row int, col int, oldGrid *ConwayGrid) {
	aliveNeighbours := oldGrid.countAliveNeighbours(row, col)
	if aliveNeighbours == 3 && oldGrid.data[row][col] == Dead {
		cg.data[row][col] = Alive
	}
	if aliveNeighbours != 3 && aliveNeighbours != 2 && oldGrid.data[row][col] == Alive {
		cg.data[row][col] = Dead
	}
}

func (cg ConwayGrid) countAliveNeighbours(row int, col int) int {
	neighbours := [8][2]int{
		{row - 1, col - 1},
		{row - 1, col},
		{row - 1, col + 1},
		{row, col - 1},
		{row, col + 1},
		{row + 1, col - 1},
		{row + 1, col},
		{row + 1, col + 1},
	}
	aliveNeighbours := 0
	for _, neighbour := range neighbours {
		r := neighbour[0]
		c := neighbour[1]
		if r >= 0 && r < len(cg.data) && c >= 0 && c < len(cg.data[0]) {
			if cg.isAlive(r, c) {
				aliveNeighbours += 1
			}
		}
	}
	return aliveNeighbours
}

func (cg ConwayGrid) isAlive(row int, col int) bool {
	return cg.data[row][col] == Alive
}

func NewGridFromFile(path string) *ConwayGrid {
	gridFile, err := os.Open(path)
	if err != nil {
		log.Fatal(err)
	}
	data, err := ioutil.ReadAll(gridFile)
	if err != nil {
		log.Fatal(err)
	}
	cellStates := make([][]CellState, 0)
	rows := strings.Split(string(data), "\n")
	for i, row := range rows {
		cellStates = append(cellStates, make([]CellState, 0))
		cells := strings.Split(row, " ")
		for _, cell := range cells {
			cellInt, err := strconv.Atoi(cell)
			if err != nil {
				log.Fatal(err)
			}
			if cellInt == 0 {
				cellStates[i] = append(cellStates[i], Dead)
			} else if cellInt == 1 {
				cellStates[i] = append(cellStates[i], Alive)
			}
		}
	}
	height := uint(len(cellStates))
	width := uint(len(cellStates[0]))
	grid := ConwayGrid{width, height, cellStates}
	return &grid
}

File: kotlin/src/main/kotlin/Conway.kt
import java.io.File

enum class CellState(private val display: String) {
    ALIVE("1"), DEAD("0");

    override fun toString(): String {
        return display
    }
}

fun newConwayGameFromFile(filename: String): Conway {
    val grid = mutableListOf<MutableList<CellState>>()
    File(filename).forEachLine {
        val row = mutableListOf<CellState>()
        it.split(" ").forEach { cell ->
            when (cell) {
                "0" -> row.add(CellState.DEAD)
                "1" -> row.add(CellState.ALIVE)
                else -> throw IllegalArgumentException(cell)
            }
        }
        grid.add(row)
    }
    return Conway(grid[0].size, grid.size, grid)
}

class Conway(private val width: Int, private val height: Int, private val grid: MutableList<MutableList<CellState>>) {

    fun step() {
        val toUpdate = mutableListOf<Pair<Pair<Int, Int>, CellState>>()
        for (i in 0 until height) {
            for (j in 0 until width) {
                val aliveNeighbours = countAliveNeighbours(i, j)
                if (aliveNeighbours == 3 && grid[i][j] == CellState.DEAD) {
                    toUpdate.add(Pair(Pair(i, j), CellState.ALIVE))
                }
                if (aliveNeighbours != 3 && aliveNeighbours != 2 && grid[i][j] == CellState.ALIVE) {
                    toUpdate.add(Pair(Pair(i, j), CellState.DEAD))
                }
            }
        }
        updateCellsAfterStep(toUpdate)
    }

    private fun updateCellsAfterStep(cellsToUpdate: List<Pair<Pair<Int, Int>, CellState>>) {
        for ((cell, cellState) in cellsToUpdate) {
            grid[cell.first][cell.second] = cellState
        }
    }

    private fun isCellAlive(i: Int, j: Int): Boolean {
        return grid[i][j] == CellState.ALIVE
    }

    private fun countAliveNeighbours(i: Int, j: Int): Int {
        val neighbours = listOf(
            Pair(i - 1, j - 1),
            Pair(i - 1, j),
            Pair(i - 1, j + 1),
            Pair(i, j - 1),
            Pair(i, j + 1),
            Pair(i + 1, j - 1),
            Pair(i + 1, j),
            Pair(i + 1, j + 1)
        )
        var aliveNeighbours = 0
        for ((r, c) in neighbours) {
            if (r >= 0 && r < grid.size && c >= 0 && c < grid[0].size && isCellAlive(r, c)) {
                aliveNeighbours += 1
            }
        }
        return aliveNeighbours
    }

    override fun toString(): String {
        var str = ""
        for (line in grid) {
            for (cell in line) {
                str += "$cell "
            }
            str += "\n"
        }
        return str
    }
}